drivers/android/transaction.rs

```diff
@@ -183,6 +250,24 @@ impl DeliverToRead for Transaction {
         // here on out.
         send_failed_reply.dismiss();
 
+        // Commit all files.
+        {
+            let mut it = file_list.cursor_front_mut();
+            while let Some(file_info) = it.current() {
+                if let Some(reservation) = file_info.reservation.take() {
+                    if let Some(file) = file_info.file.take() {
+                        reservation.commit(file);
+                    }
+                }
+
+                it.move_next();
+            }
+        }
+
+        // When `drop` is called, we don't want the allocation to be freed because it is now the
+        // user's reponsibility to free it.
+        self.free_allocation.store(false, Ordering::Relaxed);
```

Comment:
If using `Ordering::Relaxed`, is the following scenario possible?
1. a thread calls `do_work()` and `free_allocation` is set to `false`
2. another thread calls `Drop`, doesn't "see" that `free_allocation` is `false` yet, frees the allocation
3. user frees the allocation again?

Reply:
It's not possible because `Arc` prevents that. In particular, all ref count decrements have release semantics, which are paired with an acquire right before `drop`. See https://doc.rust-lang.org/src/alloc/sync.rs.html#1540.

Reply:
That definitely sounds legit :) It may be hard to know for most people looking at the code? Maybe document this in the code at some point in the future? Idk, I'm sure you've got more important things on your plate in Binder.

Reply:
Fair point. I'll add a comment that describes this.

Reply:
This got me thinking. Currently in kernel C, there are lots of variables that convey information from one thread to another - like `free_allocation` here. But many of them don't need to be protected by a synchronization primitive. That's because many kernel calls issue explicit fences. E.g. when you're waiting on a flag with `wait_event_interruptible()`, that function is guaranteed to have issued a fence when it returns, so the int or bool flag you're waiting for can never be out of date.

But in Rust, we can't make that optimization? We always need the non-zero cost synchronization primitive, even if we can logically deduce things will work across threads?

Anyways, this is too optimize-ey to worry about at this point IMHO.

Reply:
`load/store` with `Ordering::Relaxed` doesn't have overhead though. There is a nice article about what's actually generated: https://www.cl.cam.ac.uk/~pes20/cpp/cpp0xmappings.html. `load/store` with `Ordering::Relaxed` are basically just normal load/store with additional tear-free guarantee (even though normal load/store are *de facto* tear-free).

Reply:
Interesting!! Is `Relaxed` equivalent to a normal access in a structure?

Reply:
> Interesting!! Is `Relaxed` equivalent to a normal access in a structure?

If you only access it once :)

If you access an atomic integer multiple times with `Relaxed` ordering the compiler will need to generate multiple load instructions (since its value may change), while for a non-atomic integer Rust compiler is free to assume that value wouldn't change (even if there is a function-call in between two accesses).

Reply:
That's extremely interesting, thank you Gary !

Reply:
> `load/store` with `Ordering::Relaxed` doesn't have overhead though.

While this is generally true, it isn't _always_ true. Even on architectures that offer non-tearing word-sized ops, if the type is bigger than the word size, we may need more expensive instructions. For example, on 32-bit x86, if you try to atomically read/write a 64-bit integer, it won't be a regular load/store.

But to Sven's original point, I tend to agree that Rust imposes restrictions that may make the code less efficient than C. We need to look into those carefully as I think a lot of them can be exposed safely in Rust, however, some of them are inherently unsafe (i.e., they rely on the developer following rules that cannot be enforced by the compiler) -- for these I think we'll have to do as discussed in another bug/issue, provide a combination of: safe alternative with the extra cost as the price for safety, and/or an unsafe optimal alternative. If we offer both, we let the developer choose according to their needs.